File: backend/src/main/java/com/waffle/shattlebus/backend/searching/SearchHangul.java
package com.waffle.shattlebus.backend.searching;

import java.security.Key;

public class SearchHangul {
    private static final char HANGUL_BEGIN_UNICODE = 44032; // 가
    private static final char HANGUL_LAST_UNICODE = 55203; // 힣
    private static final char HANGUL_BASE_UNIT = 588; //각자음 마다 가지는 글자수
    static String text = "가깋낗닣딯띻맇밓빟삫싷잏짛찧칳킿팋핗힣";
    static String INITIAL = "ㄱㄲㄴㄷㄸㄹㅁㅂㅃㅅㅇㅈㅉㅊㅋㅌㅍㅎ";

    static char[] textArr = text.toCharArray();
    static char[] INITIAL_SOUND = INITIAL.toCharArray();

    /**
     * 해당 문자가 INITIAL_SOUND인지 검사.
     *
     * @param searchar
     * @return
     */
    private static boolean isInitialSound(char searchar){
        for(char c:INITIAL_SOUND){
            if(c == searchar){
                return true;
            }
        }
        return false;
    }

    /**
     * 해당 문자의 자음을 얻는다.
     *
     * @param c 검사할 문자
     * @return
     */
    private static char getInitialSound(char c) {
        int hanBegin = (c - HANGUL_BEGIN_UNICODE);
        int index = hanBegin / HANGUL_BASE_UNIT;
        return INITIAL_SOUND[index];
    }

    /**
     * 해당 문자가 한글인지 검사
     *
     * @param c 문자 하나
     * @return
     */
    private static boolean isHangul(char c) {
        return HANGUL_BEGIN_UNICODE <= c && c <= HANGUL_LAST_UNICODE;
    }

    /**
     * 문자 하나의 검색 범위를 찾음
     *
     * @param targetChar char 하나
     * @return endChar
     */
    static char findEndChar(char targetChar){
        char endChar = '0';

        // 입력값의 끝 문자가 ㄱ...ㅎ 인 경우
        if(isInitialSound(targetChar)) {
            for(int i = 0; i < INITIAL_SOUND.length; i++) {
                if(targetChar == INITIAL_SOUND[i]) {
                    endChar = textArr[i + 1];
                }
            }
        } else {
            // 입력값의 끝 문자가 가...힣 인 경우
            for(int i = 0; i < text.length(); i++) {
                if(targetChar >= textArr[i] && targetChar < textArr[i + 1]) {
                    if((textArr[i+1]-targetChar)%28!=27){ // 종성이 있는 경우
                        endChar = targetChar;
                    }
                    else{ // 종성이 없는 경우
                        endChar = (char) (textArr[i+1]-((textArr[i+1]-targetChar)/28)*28);
                    }
                }
            }
        }
        return endChar;
    }

    /**
     * word에 keyword가 속하는지 검사
     *
     * @param word
     * @param keyword
     * @return boolean
     */
    public static int compWord(String word, String keyword){
        int wlen = word.length(), kwlen = keyword.length();
        for(int i = 0; i <= (wlen - kwlen); i++){
            boolean flag = true;
            for(int j = 0; j < kwlen; j++) {
                char a = word.charAt(i+j);
                if (keyword.charAt(j)>a || a>findEndChar(keyword.charAt(j))) flag=false;
            }
            if(flag) return i+kwlen;
        }
        return -1;
    }
}

File: api/src/main/kotlin/com/wafflestudio/shattlebus/search/SearchHangul.kt
package com.wafflestudio.shattlebus.search

object SearchHangul {
    private const val HANGUL_BEGIN_UNICODE = 44032 // 가
        .toChar()
    private const val HANGUL_LAST_UNICODE = 55203 // 힣
        .toChar()
    private const val HANGUL_BASE_UNIT = 588 // 각자음 마다 가지는 글자수
        .toChar()
    var text = "가깋낗닣딯띻맇밓빟삫싷잏짛찧칳킿팋핗힣"
    var INITIAL = "ㄱㄲㄴㄷㄸㄹㅁㅂㅃㅅㅇㅈㅉㅊㅋㅌㅍㅎ"
    var INITIAL_BEGIN = "가까나다따라마바빠사아자짜차카타파하"
    var textArr = text.toCharArray()
    var INITIAL_SOUND = INITIAL.toCharArray()
    var INITIAL_BEGIN_SOUND = INITIAL_BEGIN.toCharArray()

    /**
     * 해당 문자가 INITIAL_SOUND인지 검사.
     *
     * @param searchar
     * @return
     */
    private fun isInitialSound(searchar: Char): Boolean {
        for (c in INITIAL_SOUND) {
            if (c == searchar) {
                return true
            }
        }
        return false
    }

    /**
     * 해당 문자가 한글인지 검사
     *
     * @param c 문자 하나
     * @return
     */
    private fun isHangul(c: Char): Boolean {
        return isInitialSound(c) || HANGUL_BEGIN_UNICODE <= c && c <= HANGUL_LAST_UNICODE
    }

    /**
     * 문자 하나의 검색 범위를 찾음
     *
     * @param searchar char 하나
     * @return beginChar
     */
    fun findBeginChar(searchar: Char): Char {
        for (i in INITIAL_SOUND.indices) {
            if (INITIAL_SOUND[i] == searchar) {
                return INITIAL_BEGIN_SOUND[i]
            }
        }
        return searchar
    }

    /**
     * 문자 하나의 검색 범위를 찾음
     *
     * @param targetChar char 하나
     * @return endChar
     */
    fun findEndChar(targetChar: Char): Char {
        var endChar = '0'

        // 입력값의 끝 문자가 ㄱ...ㅎ 인 경우
        if (isInitialSound(targetChar)) {
            for (i in INITIAL_SOUND.indices) {
                if (targetChar == INITIAL_SOUND[i]) {
                    endChar = textArr[i + 1]
                }
            }
        } else {
            // 입력값의 끝 문자가 가...힣 인 경우
            for (i in 0 until text.length) {
                if (targetChar >= textArr[i] && targetChar < textArr[i + 1]) {
                    endChar = if ((textArr[i + 1] - targetChar) % 28 != 27) { // 종성이 있는 경우
                        targetChar
                    } else { // 종성이 없는 경우
                        textArr[i + 1] - (textArr[i + 1] - targetChar) / 28 * 28
                    }
                }
            }
        }
        return endChar
    }

    /**
     * word에 keyword가 속하는지 검사
     *
     * @param word
     * @param keyword
     * @return boolean
     */
    fun compWord(word: String, keyword: String): Int {
        val wlen = word.length
        val kwlen = keyword.length
        for (i in 0..wlen - kwlen) {
            var flag = true
            for (j in 0 until kwlen) {
                val a = word[i + j]
                val b = keyword[j]
                if (isHangul(a) && isHangul(b) && (findBeginChar(b) > a || a > findEndChar(b))) flag = false
                if ((!isHangul(a) || !isHangul(b)) && a.uppercaseChar() != b.uppercaseChar()) flag = false
            }
            if (flag) return i + kwlen
        }
        return -1
    }
}
